src/rp2/plugin/accounting_method/hifo.py

```diff
@@ -0,0 +1,188 @@
+        taxable_event_amount: RP2Decimal,
+        acquired_lot_amount: RP2Decimal,
+    ) -> TaxableEventAndAcquiredLot:
+        new_acquired_lot: Optional[InTransaction] = acquired_lot
+        new_acquired_lot_amount: RP2Decimal = acquired_lot_amount - taxable_event_amount if acquired_lot is not None else ZERO
+
+        try:
+            new_taxable_event: AbstractTransaction = next(self.__taxable_event_iterator)
+        except StopIteration:
+            raise TaxableEventsExhaustedException() from None
+        new_taxable_event_amount: RP2Decimal = new_taxable_event.crypto_balance_change
+
+        # If the new taxable event has different year than the acquired lot (and it's not earn-typed), also get a new acquired lot from the new year
+        if taxable_event and taxable_event.timestamp < new_taxable_event.timestamp:
+            if acquired_lot:
+                # Cache old-year acquired_lot amount
+                self.__acquired_lot_2_partial_amount[acquired_lot] = new_acquired_lot_amount
+            (_, new_acquired_lot, _, new_acquired_lot_amount) = self.get_acquired_lot_for_taxable_event(
+                new_taxable_event, acquired_lot, new_taxable_event_amount, new_acquired_lot_amount
+            )
+
+        return TaxableEventAndAcquiredLot(
+            taxable_event=new_taxable_event,
+            acquired_lot=new_acquired_lot,
+            taxable_event_amount=new_taxable_event_amount,
+            acquired_lot_amount=new_acquired_lot_amount,
+        )
+
+    def get_acquired_lot_for_taxable_event(
+        self, taxable_event: AbstractTransaction, acquired_lot: Optional[InTransaction], taxable_event_amount: RP2Decimal, acquired_lot_amount: RP2Decimal
+    ) -> TaxableEventAndAcquiredLot:
+        # This while loop makes the algorithm's complexity O(n^2), where n is the number of acquired lots. Non-trivial
+        # optimizations are possible using different data structures (and likely with some space/time tradeoff)
+        new_taxable_event_amount: RP2Decimal = taxable_event_amount - acquired_lot_amount
+        index = len(self.__spot_price_list) - 1
+        while index >= 0:
+
+            acquired_lot_list: List[InTransaction] = self.__acquired_lot_list
+            first_lot: Optional[AcquiredLotAndIndex] = self.__acquired_lot_avl.find_max_value_less_than(
+                f"{self._get_avl_node_key_with_max_disambiguator(self.__spot_price_list[index])}"
+            )
+
+            if first_lot is not None and first_lot.acquired_lot.timestamp < taxable_event.timestamp:
+                acquired_lot_index: int = first_lot.index
+                if first_lot.acquired_lot != acquired_lot_list[acquired_lot_index]:
+                    raise Exception("Internal error: acquired_lot incongruence in HIFO accounting logic")
+                acquired_lot_and_amount: Optional[AcquiredLotAndAmount] = self.seek_acquired_lot(reversed(acquired_lot_list[: acquired_lot_index + 1]))
```

Comment:
If we consider the loop inside seek_acquired_lot(), the complexity of this algorithm is actually O(n^3) (where n is the number of acquired lots). The same concern applies to the LIFO plugin I wrote earlier (and which has an implementation similar to this), although LIFO's per-year bucketization, helps a little bit (it's still O(n^3) but n is the number of acquired lots in one given year).

This is fine for a first implementation, and it may be good enough in most reasonable cases. We'll have to get feedback from users. If it proves to be too slow for people with large numbers of transactions, we'll need to think of redesign this (and LIFO) with optimization in mind.